### PR TITLE
Make resolve_parameters/is_parameterized recurse on tuples and lists

### PR DESCRIPTION
**Pull request by bryano**

### CODE REVIEW

File: cirq/protocols/resolve_parameters.py [hunk 1]
```diff
@@ -88,10 +89,13 @@ def resolve_parameters(
     param_resolver = ParamResolver(param_resolver)
     if isinstance(val, sympy.Basic):
         return param_resolver.value_of(val)
+    if isinstance(val, Sequence):
+        return type(val)(resolve_parameters(e, param_resolver) for e in val)
```
Comment (Strilanc) [inline cirq/protocols/resolve_parameters.py]: This is not safe. Sequence types are not guaranteed to have a constructor that takes a generator. I think all a type has to do to be a sequence is have `__len__` and `__getitem__`.

I'd recommend having an explicit list of types we support, like just tuple and list and dict.

Comment (bryano) [reply cirq/protocols/resolve_parameters.py]: Okay, I changed it to list and tuples explicitly.

Is it clear to you how dicts should be resolved (i.e., just the keys, just the values, or both)? I think it's ambiguous enough that we should wait for a use case.

Comment (Strilanc) [reply cirq/protocols/resolve_parameters.py]: I agree.

File: cirq/protocols/resolve_parameters.py [hunk 2]
```diff
@@ -88,10 +89,13 @@ def resolve_parameters(
...
+    if isinstance(val, (list, tuple)):
+        return type(val)(resolve_parameters(e, param_resolver) for e in val)
 
     getter = getattr(val, '_resolve_parameters_', None)
     result = NotImplemented if getter is None else getter(param_resolver)
 
     if result is not NotImplemented:
         return result
-    return val
+    else:
```
Comment (Strilanc) [inline cirq/protocols/resolve_parameters.py]: Victory just went around getting rid of this "return else return" thing.

**Comment by bryano** [comment]
Added `is_parameterized`.

File: cirq/protocols/resolve_parameters_test.py [hunk 3]
```diff
@@ -58,6 +58,18 @@ def _resolve_parameters_(self, resolver: ParamResolver):
     assert cirq.resolve_parameters(SimpleParameterSwitch('a'), d).parameter == 0
     assert cirq.resolve_parameters(sympy.Symbol('a'), r) == 0
 
+    a, b, c = tuple(sympy.Symbol(l) for l in 'abc')
+    x, y, z = 0, 4, 7
+    resolver = {a: x, b: y, c: z}
+    assert cirq.resolve_parameters((a, b, c), resolver) == (x, y, z)
```
Comment (Strilanc) [inline cirq/protocols/resolve_parameters_test.py]: Check the empty list corner case, and also that a list with no parameters is returned unchanged.

**Pull request merged**